README.md
# popup_widget

A pop-up bubble text component, lightweight, easy to use.

## Getting Started

![img1](popup_one.png) 

![img2](popup_two.png) 

## Example

```
PopupWidget(
    child: Icon(Icons.help_outline, size: 16, color: Colors.blue),
    text: "Lord, I'm one, Lord",
    contentSize: Offset(16, 16),
);
```

## Contact
Github：[Github](https://github.com/linminjing888/popup_widget)  
掘金：[minjing_lin](https://juejin.cn/user/1961184474981688)



CHANGELOG.md
## [0.1.3] - 2021.02.02

* Error
## [0.1.2] - 2021.02.02

* README、Library
## [0.1.1] - 2021.02.02

* LICENSE、Example
## [0.1.0] - 2021.02.02

* Basic bubble pop-up function



test/popup_widget_test.dart
import 'package:flutter_test/flutter_test.dart';

// import 'package:popup_widget/popup_widget.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    // expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/popup_tip.dart
import 'package:flutter/material.dart';
import 'package:popup_widget/src/popup_triangle_painter.dart';

/// 气泡UI
class PopupTip extends StatelessWidget {
  final Offset positionXy;
  final String text;
  final Color textColor;
  final Color bgColor;
  final bool isShadow;
  final Color shadowColor;

  PopupTip({
    this.positionXy = Offset.zero,
    this.text = "",
    this.textColor = Colors.white,
    this.bgColor = const Color.fromRGBO(75, 75, 75, 1.0),
    this.isShadow = false,
    this.shadowColor = Colors.grey,
  });

  @override
  Widget build(BuildContext context) {
    double screenW = MediaQuery.of(context).size.width;
    Size sz = _getTextSize(text, TextStyle(fontSize: 13));
    double textWidth = sz.width;

    // double ww = 20;
    // double left = ww; // 左右默认间距
    // double right = ww;
    // print("$textWidth-${positionXy.dx}-$screenW");
    // if (positionXy.dx + textWidth >= screenW) {
    //   left = screenW - right - textWidth - 40; // 40是文字与容器的左右距离和
    //   print("$left");
    // } else {
    //   right = screenW - left - textWidth - 40;
    // }

    double padding = 20;
    double minX = positionXy.dx - textWidth / 2 - padding;
    double maxX = positionXy.dx + textWidth / 2 + padding;
    double left = minX;
    double right = screenW - maxX;
    if (right < padding) {
      right = padding;
      left = screenW - right - textWidth - padding * 2;
      if (left < padding) {
        left = padding;
      }
    }
    if (left < padding) {
      left = padding;
      right = screenW - left - textWidth - padding * 2;
    }

    double triangleW = padding;
    double triangleL = positionXy.dx - triangleW / 2;
    if (triangleL < triangleW + 10) {
      // print("===L$triangleL");
      triangleL = triangleW + 10;
    }

    return new Material(
      type: MaterialType.transparency, // Material类型设置
      child: new GestureDetector(
        child: new Stack(
          children: <Widget>[
            Container(
              // 设置一个容器组件，是整屏幕的。
              width: MediaQuery.of(context).size.width,
              height: MediaQuery.of(context).size.height,
              color: Colors.transparent,
            ),
            Positioned(
              top: positionXy.dy + 15 - triangleW,
              left: triangleL,
              width: triangleW,
              height: triangleW,
              child: CustomPaint(
                painter: PopTrianglePainter(color: bgColor, isShadow: isShadow),
              ),
            ),
            Positioned(
              child: Container(
                padding:
                    EdgeInsets.symmetric(horizontal: padding, vertical: 10),
                decoration: BoxDecoration(
                  color: bgColor,
                  borderRadius: BorderRadius.all(Radius.circular(5.0)),
                  boxShadow: [
                    if (isShadow)
                      BoxShadow(
                          offset: Offset(.5, 3),
                          color: shadowColor,
                          blurRadius: 3)
                  ],
                ),
                child: Text(text,
                    maxLines: 1,
                    textAlign: TextAlign.center,
                    style: TextStyle(color: textColor, fontSize: 13)),
              ),
              top: positionXy.dy + 15,
              left: left,
              right: right,
            )
          ],
        ),
        onTap: () => Navigator.of(context).pop(), //点击空白处直接返回
      ),
    );
  }

  // 获取字体Size
  Size _getTextSize(String text, TextStyle style) {
    final TextPainter textPainter = TextPainter(
        text: TextSpan(text: text, style: style),
        maxLines: 1,
        textDirection: TextDirection.ltr)
      ..layout(minWidth: 0, maxWidth: double.infinity);
    return textPainter.size;
  }
}


lib/src/pop_widget.dart
/*
 * @Author: lin minjing
 * @version: 0.1.0
 * @Date: 2021-02-01 16:30:19
 * @Descripttion: 气泡弹出
 */
import 'package:flutter/material.dart';
import 'package:popup_widget/src/pop_route.dart';
import 'package:popup_widget/src/popup_tip.dart';

class PopupWidget extends StatelessWidget {
  /// 子组件
  final Widget child;

  /// 文本字符串
  final String text;

  /// 文本颜色
  final Color textColor;

  /// 背景色
  final Color bgColor;

  /// 是否显示阴影
  final bool isShadow;

  /// 阴影颜色
  final Color shadowColor;

  /// child 尺寸帮助精确定位，不传的话，位置会一直变动
  final Offset contentSize;

  PopupWidget({
    this.child,
    @required this.text,
    this.textColor = Colors.white,
    this.bgColor = const Color.fromRGBO(75, 75, 75, 1.0),
    this.isShadow = false,
    this.shadowColor = Colors.grey,
    this.contentSize = Offset.zero,
  });

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      child: child,
      onTapDown: (detail) {
        print("${detail.localPosition}-${detail.globalPosition}");

        double xx = detail.globalPosition.dx - detail.localPosition.dx;
        double yy = detail.globalPosition.dy - detail.localPosition.dy;

        // 使用路由跳转方式
        Navigator.push(
          context,
          new PopRoute(
            child: PopupTip(
              positionXy: Offset(xx + contentSize.dx / 2, yy + contentSize.dy),
              text: text,
              textColor: textColor,
              bgColor: bgColor,
              isShadow: true,
              shadowColor: shadowColor,
            ),
          ),
        );
      },
    );
  }
}


lib/popup_triangle_painter.dart
import 'package:flutter/material.dart';

/// 绘制三角形
class PopTrianglePainter extends CustomPainter {
  /// 填充颜色
  Color color;

  /// 是否显示阴影
  final bool isShadow;

  PopTrianglePainter({
    this.color = const Color.fromRGBO(75, 75, 75, 1.0),
    this.isShadow = false,
  });

  @override
  void paint(Canvas canvas, Size size) {
    Paint _paint = Paint()
      ..strokeWidth = 1.0 //线宽
      ..color = color;

    Path _path = Path();

    final baseX = size.width * 0.5;
    final baseY = size.height * 0.5;
    //起点
    _path.moveTo(0, baseY * 2);
    _path.lineTo(baseX, 0.4 * baseY);
    _path.lineTo(baseX * 2, baseY * 2);
    canvas.drawPath(_path, _paint);
    if (isShadow == true) {
      canvas.drawShadow(_path, Colors.black, 0.5, false);
    }
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:popup_widget/popup_widget.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'PopupWidget Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('PopupWidget'),
      ),
      body: Center(
        child: RichText(
          textAlign: TextAlign.right,
          maxLines: 1,
          text: TextSpan(
            style: TextStyle(color: Color(0xFF333333), fontSize: 15),
            children: [
              TextSpan(text: "Lord"),
              WidgetSpan(
                child: PopupWidget(
                  child: Icon(Icons.help_outline, size: 16, color: Colors.blue),
                  text: "Lord, I'm one, Lord",
                  contentSize: Offset(16, 16),
                ),
              ),
              TextSpan(text: "I'm five hundred miles away from home"),
              WidgetSpan(
                child: PopupWidget(
                  child: Icon(Icons.mood, size: 16, color: Colors.blue),
                  text: "away from home, away from home",
                  bgColor: Colors.white,
                  textColor: Colors.blue,
                  isShadow: true,
                  contentSize: Offset(16, 16),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}
